File: 2020/day_06/day_06.py
with open('in') as f:
    data = list(map(lambda x: x.strip(), f.readlines()))


def part_one():
    count = 0
    group = set()

    for line in data:
        if line == '':
            count += len(group)
            group = set()
            continue

        group.update(line)

    count += len(group)
    return count


def part_two():
    count = 0
    group = set()
    empty = True

    for line in data:
        if line == '':
            count += len(group)
            group = set()
            empty = True
            continue

        if empty:
            group.update(line)
            empty = False
        else:
            group.intersection_update(line)

    count += len(group)
    return count


if __name__ == '__main__':
    print(f'Part One: {part_one()}')
    print(f'Part Two: {part_two()}')

File: 2020/day_06/day_06.rs
use std::fs;
use std::collections::HashSet;

fn main() {
    let f = fs::read_to_string("in").unwrap();
    let data: Vec<&str> = f.lines().map(|x| x.trim()).collect();

    println!("Part One: {}", part_one(&data));
    println!("Part Two: {}", part_two(&data));
}

fn part_one(data: &Vec<&str>) -> u32 {
    let mut count = 0;
    
    let mut current = HashSet::new();
    for line in data {
        if *line == "" {
            count += current.len();
            current.clear();
        } else {
            for c in line.chars() {
                current.insert(c);
            }
        }
    }

    (count + current.len()) as u32
}

fn part_two(data: &Vec<&str>) -> u32 {
    let mut count = 0;

    let mut current = HashSet::new();
    let mut empty = true;
    for line in data {
        if *line == "" {
            count += current.len();
            current.clear();
            empty = true;
        } else {
            let mut person = HashSet::new();
            for c in line.chars() {
                person.insert(c);
            }

            if empty {
                current = current.union(&person).map(|x| *x).collect();
                empty = false;
            } else {
                current = current.intersection(&person).map(|x| *x).collect();
            }
        }
    }

    (count + current.len()) as u32
}